docs/proposals/deep-links.md

```diff
@@ -0,0 +1,153 @@
+
+
+## Proposal
+
+The configuration for Deep Links will be present in the `argocd-cm`, we will add new `<location>.links` fields in the
+cm to list all the deep links that will be displayed in the provided location. The possible values for `<location>`
+currently are :
+- `project` : all links under this field will show up in the project tab in the Argo CD UI.
+- `application` : all links under this field will show up in the application summary tab.
+- `resource` : all links under this field will show up in the individual resource(deployments, pods, services etc.)
+  summary tab.
+
+Each link in the list has three sub-fields :
+1. `title` : title/tag that will be displayed in the UI corresponding to that link
+2. `url` : the actual URL where the deep link will redirect to, this field can be templated to use data from the
+   application, project, resource objects (depending on where it is located) or from argocd-secret (only fields prefixed
+   with `links.`)
+3. `description` (optional) : a descriptiopn for what the deep link is about
+4. `icon.class` (optional) : a font-awesome icon class to be used when displaying the links in dropdown menus.
+5. `if` (optional) : a conditional statement that results in either `true` or `false`, it also has access to the same
+   data as the`url` field. If the condition resolves to `true` the deep link will be displayed else it will be hidden. If
+   the field is omitted by default the deep links will be displayed.
+
+
+An example `argocd-cm.yaml` file with deep links and its variations :
+
+```yaml
+data:
+  # project level links
+  project.links: |
+    - url: https://myaudit-system.com?project={{proj.metadata.name}}
+      title: Audit
+  # application level links
+  application.links: |
+    - url: https://mycompany.splunk.com?search={{app.spec.destination.namespace}}
+      title: Splunk
+    # reference secrets from `argocd-secret`
+    - url: https://mycompany.splunk.com?token={{secret.something}}
+      title: Splunk
+    # conditionally show link e.g. for specific project
+    - url: https://mycompany.splunk.com?token={{secret.something}}
+      title: Splunk
+      if: app.spec.proj == "abc" # https://github.com/antonmedv/expr
+    # conditionally show link e.g. for specific project
+    - url: https://{{project.metadata.annotations.splunkhost}}?token={{secret.something}}
+      title: Splunk
+      if: project.metadata.annotations.splunkhost
+    
+  # resource level links
+  resource.links: |
+    - url: https://mycompany.splunk.com?search={{res.metadata.namespace}}
+      title: Splunk
+      if: res.kind == "Pod" || res.kind == "Deployment"
+
+```
+
+Example `argocd-secret` resource with deep link secrets :
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: argocd-secret
+  namespace: argocd
+  labels:
+    app.kubernetes.io/name: argocd-secret
+    app.kubernetes.io/part-of: argocd
+type: Opaque
+stringData:
+  links.something: mySecret
```

Comment:
What are the advantages to storing link variables in the argocd-secret? I worry about having additional code accessing the secret that contains `server.secretKey`. I also worry about encouraging more edits of a secret which is critical to Argo CD's functioning.

Reply:
agreed, we shouldn't store extraneous things in argocd-secret

Reply:
Good point. I'm with you here, in that I'd encourage the implementation without referring to any secret at all.